### ISSUE TITLE
OSC works fine. Maybe too much

### ISSUE THREAD

**Issue by lowgos** [colinbdclark/osc.js#145]
Hello everyone,

I'm currently working on an API to control an amphitheater and to remotely control our Behringer x32 Midas sound panel. I'm using `osc.js` in a Dockerized application living on a Debian machine.

I'm relatively new to UDP/Serial management and I am certain that the problem is caused by my work and not yours.

# The problem

Everything works fine and I didn't have any problem to setup my app.
I use the UDP module to connect and send commands.

The problem is that after a while, the x32 sound table isn't doing much when I try to remotely control it. Commands are sent, my API doesn't crash, no Exceptions etc

But I think my code is too basic and bad. I mean, I surely don't use connection functions as I should if I had to guess. And because I didn't really understand how to catch errors I think I don't catch errors... Even worse I think I don't get any feedback from the x32.

Bear with me, I know that everything is bad in the previous statement. I don't do what I should 😅I'm using a Class that a past co-worker used to use. I understand it globally but I can't say I fully understand how events/methods of `osc.js` work. And I'm sorry if I missed it after my researches but I didn't find a good documentation to try more myself, hence the S.O.S.

# Example

This is how I initialise my Class with `osc.js` and its UDP module

``` javascript
let { UDPPort } = require('osc')

// object storing the data coming from the x32
let levels = {}

let udpPort = new UDPPort({
  localAddress: '0.0.0.0',
  localPort: 10023,
  remoteAddress: '10.31.1.184',
  remotePort: 10023,
  broadcast: true
})

// A class to better access my methods
class X32 {

  // after analysing this file, I noticed this is never used. And I think it's the root of my problem
  closeOSC () {
    udpPort.close()
  }

  openOSC () {
    udpPort.open()
  }

  // function fired to get x32 state at connection
  // I'm not even sure that it's what I read after connection (see note1 below)
  messageOSC () {
    udpPort.on('message', function (oscMsg) {
      levels[oscMsg.address] = parseFloat(oscMsg.args)
    })
  }

  // I thought it was to get feedback errors but I think it only fires at first connection
  errorOSC () {
    udpPort.on('error', function (err) {
      console.error(err)
    })
  }

  readyOSC () {
    udpPort.on('ready', function () {})
  }

  loadOSC () {
    this.readyOSC()
    this.openOSC()
    this.messageOSC()
    this.errorOSC()
  }

  getFaderValue (fader) {
    this.send('/ch/' + fader + '/mix/fader')
  }

  getBusFaderValue (fader) {
    this.send('/bus/' + fader + '/mix/fader')
  }

  constructor () {
    this.loadOSC()

    // note1: I'm not even sure why my coworker sent those commands...
    setInterval(() => {
      for (let index = 1; index <= 32; index++) {
        this.getFaderValue(this.numberFormat(index))
      }
      for (let index = 1; index <= 16; index++) {
        this.getBusFaderValue(this.numberFormat(index))
      }
    }, 100)
  }

  // method I use to send commands to the x32
  send (address, args) {
    udpPort.send({
      address: address,
      args: args
    })
  }

```

And here is how I send a command to set a value (level) on an input fader

``` javascript
setChannelFader (fader, level) {
    this.send('/ch/' + fader + '/mix/fader', [level])
}
```

I think not closing the connection after all this mess provoke problems after a while (one or two hour suffice to freeze remote control on my x32, I hope it's not the device itself that breaks !).
I would like to check this hypothesis with you but I don't really know :

- How to get feedback data on how the query goes
- How to get feedback errors
- If using the UDP module is the best idea. Wouldn't TCP work ?

I'm really sorry. I know it's not the usual thing to do to come in issue to find tech support and that I shouldn't use code I don't understand but I really think your package is simple enough to come here for a bit of help !

# tl;dr

I use `osc.js` and it works but after a while my Behringer x32 freezes and I'm a moron who doesn't understand what he uses. Did I miss documentation ?

> Thanks for reading. I hope I don't waste your time.

**Comment by Lange** [comment]
The Behringer X32 has a bug in it which can cause it to freeze when using OSC commands. I am in the process of attempting to find a reliable repro for this bug. That work is ongoing here: https://github.com/GamesDoneQuick/x32-fuzzer

In other words: this is most likely a bug with the mixer, and not with `osc.js`. This issue can probably be closed.

**Comment by colinbdclark** [comment]
Thanks for clarifying @Lange. I hope this helps with your issue, @lowgos. I'm going to close this issue as suggested, but feel free to reopen if new information crops up.

**Comment by lowgos** [comment]
After tinkering for a bit I cleaned my code and waited for request to cleanly end.

I've also delayed some multi-request and the x32 behavior seems to follow what @Lange posted. In every way, my code could instanciate a lot of concurrent clients, making the x32 bug.

Thank you for letting me know that it's a known issue.

I've already got something reliable and working. Still it's good to know.
